jdaviz/configs/mosviz/plugins/viewers.py

```diff
@@ -88,14 +93,118 @@ def __init__(self, *args, **kwargs):
+
+        def pixel_to_world_line(pixel):
+            return line_coeffs[0] * pixel + line_coeffs[1]
+
+        def map_pixel_to_world(pixel):
+            if pixels[0] <= pixel <= pixels[-1]:
+                # interpolate directly
+                return float(self.pixel_to_world_interp(pixel))
+            else:
+                # use the line model to extrapolate
+                return pixel_to_world_line(pixel)
+
+        invert = (-1) ** sum((self.inverted_x_axis, limits[0] > limits[1]))
+        out_lims = list(map(map_pixel_to_world, limits))[::invert]
+
+        return out_lims
+
+    @cached_property
+    def world_to_pixel_interp(self):
+        spectral_axis = self.state.reference_data.get_object().spectral_axis.value
+        pixels = range(len(spectral_axis))
+        return interp1d(spectral_axis, pixels)
+
+    def world_to_pixel_limits(self, limits):
+        if not len(limits) == 2:
+            raise ValueError("limits must be length 2")
+
+        spectral_axis = self.state.reference_data.get_object().spectral_axis.value
+
+        # we'll use interpolation when possible, but also want to fit a line between
+        # the outermost edge of the data within the limits
+        line_edges_world = np.array([max((min(spectral_axis), min(limits))),
+                                     min((max(spectral_axis), max(limits)))])
+        if line_edges_world[0] > line_edges_world[1]:
+            # then the limits are entirely out of range, so use the whole range
+            # when fitting the linear approximation
+            line_edges_world = np.array([min(spectral_axis), max(spectral_axis)])
+        line_edges_pixels = self.world_to_pixel_interp(line_edges_world)
+        line_coeffs = np.polyfit(line_edges_world, line_edges_pixels, deg=1)
+
+        def world_to_pixel_line(world):
+            return line_coeffs[0] * world + line_coeffs[1]
+
+        def map_world_to_pixel(world):
+            if spectral_axis[0] <= world <= spectral_axis[-1]:
+                # interpolate directly
+                return float(self.world_to_pixel_interp(world))
+            else:
+                # use the line model to extrapolate
+                return world_to_pixel_line(world)
+
+        invert = (-1) ** sum((self.inverted_x_axis, limits[0] > limits[1]))
+        out_lims = list(map(map_world_to_pixel, limits))[::invert]
+
+        return out_lims
+
+    def _on_viewer_data_changed(self, msg):
+        if msg.viewer_reference != self.reference:
+            return
+        # clear cached properties that are based on reference data
+        for attr in ('inverted_x_axis', 'pixel_to_world_interp', 'world_to_pixel_interp'):
+            if attr in self.__dict__:
+                del self.__dict__[attr]
+        if len(self.data()):
+            self._handle_x_axis_orientation()
+
+    @cached_property
+    def inverted_x_axis(self):
+        spectral_axis = self.data()[0].spectral_axis
```

Comment:
Do we need to keep this in mind when handling reference data being removed?

Reply:
the cache clears whenever data is added to/removed from the viewer (which is probably a bit overly conservative, but it's also not expensive).  This uses the first entry (which should always be the initially loaded 2d spectrum)... we could change this to `self.state.reference_data.get_object().spectral_axis` and then make sure to clear the cache on a change to reference data as well (or perhaps instead?).

I guess there could be a scenario where the user loads a second 2d spectrum invert with respect to the first and then changes the reference from the first to the second.  We would need to think about what behavior is expected in that situation.

Reply:
Got it, just wanted to bring that up since we will be discussing this more in a week. We can keep this as is and improve it in a follow-up once we've made that decision.

Reply:
yes, I think we should also give some thought - in general - to workflows for multi-2d-spectra in specviz2d, or disable as we did in cubeviz.  But this logic should be able to easily be modified based on those decisions 🤞